### PR TITLE
rbd-mirror: image sync needs to handle snapshot size and protection status

### PR DESCRIPTION
**Pull request by dillaman**

### CODE REVIEW

File: src/tools/rbd_mirror/image_sync/SnapshotCopyRequest.cc [hunk 1]
```diff
@@ -79,31 +80,133 @@ void SnapshotCopyRequest<I>::send_snap_remove() {
...
+    librados::snap_t local_snap_id = *snap_id_it;
+    {
+      RWLock::RLocker local_snap_locker(m_local_image_ctx->snap_lock);
+      bool local_unprotected;
+      int r = m_local_image_ctx->is_snap_unprotected(local_snap_id,
+                                                     &local_unprotected);
+      if (r < 0) {
+        lderr(cct) << "failed to retrieve local snap unprotect status: "
+                   << cpp_strerror(r) << dendl;
+        finish(r);
```
Comment (trociny) [inline src/tools/rbd_mirror/image_sync/SnapshotCopyRequest.cc]: Is it safe to run finish() holding snap_lock?

Comment (dillaman) [inline src/tools/rbd_mirror/image_sync/SnapshotCopyRequest.cc]: Thx --- I'll fix

**Comment by jdurgin** [comment]
lgtm

File: src/test/Makefile-client.am [hunk 2]
```diff
@@ -317,6 +317,11 @@ noinst_HEADERS += \
 	test/librados_test_stub/TestIoCtxImpl.h
 noinst_LTLIBRARIES += librados_test_stub.la
 
+libjournal_test_mock_la_SOURCES = \
+	test/journal/mock/MockJournaler.cc
+libjournal_test_mock_la_CXXFLAGS = $(UNITTEST_CXXFLAGS)
```
Comment (dillaman) [inline src/test/Makefile-client.am]: missing noinst_HEADERS for MockJournaler.h

Comment (jdurgin) [inline src/test/Makefile-client.am]: thx, not sure why it built on my machine

Comment (dillaman) [inline src/test/Makefile-client.am]: The annoyance of the dist tarball that the github make bit doesn't use

Comment (jdurgin) [inline src/test/Makefile-client.am]: ah right, makes sense

**Pull request merged**